File: source_formatting/image_numbering.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

from source_formatting.html_parser import BaseHtmlTranslator

class ImageNumberingFormatter(BaseHtmlTranslator):
    def __init__(self):
        super().__init__()
        self.image_tags = []
        self.has_processed_text = False
        self.is_in_image = False

    def handle_starttag(self, tag, attrs):
        if tag == "img":
            new_tag = []
            self.append_start_tag(tag, attrs, to=new_tag)
            self.image_tags.append("".join(new_tag))
            self._out.append(None)
        else:
            self.append_start_tag(tag, attrs)

    def handle_endtag(self, tag):
        self.append_end_tag(tag)

    def handle_data(self, data):
        self._out.append(data)

    def format_image_ref(self, counter):
        if counter == 1 and len(self.image_tags) == 1:
            return "(*)"
        else:
            return f"({counter})"

    def before_join(self):
        if len(self.image_tags) == 0:
            return

        new_out = []
        image_counter = 0
        for piece in self._out:
            if piece:
                new_out.append(piece)
            else:
                image_ref = self.format_image_ref(image_counter + 1)
                new_out += [
                    '<span class="image-ref-container">',
                    f'<span class="image-ref-text">{image_ref}:</span>',
                    '<span class="image-ref">',
                    self.image_tags[image_counter],
                    "</span>",
                    "</span>",
                    f'<span class="image-pointer">{image_ref}</span>',
                ]
                image_counter += 1
        self._out = new_out

File: source_formatting/image_numbering.ts
import {Attributes, HtmlVisitor} from "./html_visitor";

// This is intentional - we want a value that is not identical to any other value
// eslint-disable-next-line no-new-wrappers
const placeholder = new String() as string;

export class ImageNumberingFormatter extends HtmlVisitor {
  imageTags: string[] = [];
  hasProcessedText = false;
  isInImage = false;

  shouldRun(input: string): boolean {
    return input.indexOf("<img") !== -1;
  }

  visitStartTag(tag: string, attributes: Attributes): void {
    if (tag === "img") {
      const newTag: string[] = [];
      this.appendStartTag(tag, attributes, newTag);
      this.imageTags.push(newTag.join(""));
      this._out.push(placeholder);
    } else {
      this.appendStartTag(tag, attributes);
    }
  }

  visitEndTag(tag: string): void {
    this.appendEndTag(tag);
  }

  visitText(text: string): void {
    this.appendText(text);
  }

  formatImageRef(counter: number): string {
    if (counter === 1 && this.imageTags.length === 1) {
      return "(*)";
    } else {
      return `(${counter})`;
    }
  }

  beforeJoin(): void {
    if (this.imageTags.length === 0) {
      return;
    }

    const newOut: string[] = [];
    let imageCounter = 0;

    for (const piece of this._out) {
      if (piece !== placeholder) {
        newOut.push(piece);
      } else {
        const imageRef = this.formatImageRef(imageCounter + 1);
        newOut.push(
          '<span class="image-ref-container">',
          `<span class="image-ref-text">${imageRef}:</span>`,
          '<span class="image-ref">',
          this.imageTags[imageCounter],
          "</span>",
          "</span>",
          `<span class="image-pointer">${imageRef}</span>`,
        );
        imageCounter += 1;
      }
    }

    this._out = newOut;
  }
}
